crates/sui-core/src/authority.rs

```diff
@@ -1544,28 +1612,49 @@ impl AuthorityState {
         let mut limit = limit.unwrap_or(usize::MAX);
         while limit > 0 {
             limit -= 1;
-            if let Some((cert, tx_guard)) = self.database.wal.read_one_recoverable_tx().await? {
+            if let Some((cert, commit_phase, tx_guard)) =
+                self.database.wal.read_one_recoverable_tx().await?
+            {
                 let digest = tx_guard.tx_id();
                 debug!(?digest, "replaying failed cert from log");
 
-                if tx_guard.retry_num() >= MAX_TX_RECOVERY_RETRY {
-                    // This tx will be only partially executed, however the store will be in a safe
-                    // state. We will simply never reach eventual consistency for this TX.
-                    // TODO: Should we revert the tx entirely? I'm not sure the effort is
-                    // warranted, since the only way this can happen is if we are repeatedly
-                    // failing to write to the db, in which case a revert probably won't succeed
-                    // either.
-                    error!(
-                        ?digest,
-                        "Abandoning in-progress TX after {} retries.", MAX_TX_RECOVERY_RETRY
-                    );
-                    // prevent the tx from going back into the recovery list again.
-                    tx_guard.release();
-                    continue;
-                }
+                // If we have already committed to WAL, then the tx has already been executed
+                // and will not be re-executed, therefore there is no longer a threat of poison pill.
+                // We can safely retry past the limit.
+                match commit_phase {
```

Comment:
Two reasons for this:

1. Calling `commit_and_notify_from_recovery` rather than `process_certificate` in the case where we're already in ``TransactionExecuted` phase saves us a repeated DB read of the tx output which we already got from checking the tx state. Probably not a huge improvement given that the data will likely still be in the page cache, but I felt it was worth doing

2. Currently `retry_num` is still incremented in `begin_tx` even for txes that are already executed. But once the tx is successfully executed, there is no threat of poison pill, so we should be able to skip checking the number of retries. Alternatively, I could just not increment the `retry_num` if the tx is already executed, but that's more work and technically less semantically correct IMO.

Reply:
Re 1, this code path should only affect maybe 1 out of a million txns (or less, quite likely) so all else being equal I think simpler code is more important than faster code.

Re 2, what you're saying makes sense, but are you sure it justifies the extra code? If we almost certainly won't crash in the Executed case, there's no harm in checking and incrementing the retry count one last time.

I'm happy either way though, so this is up to you.